### DISCUSSION TITLE
Using default servo pin in BTT_SKR2 for pwm spindle control

### DISCUSSION THREAD

**Discussion by krishchan007** [grblHAL/STM32F4xx#96] [General]
Hi,
Is it possible to use default servo pin (PE5) in BTT_SKR2 for pwm spindle control. This pin has TIM9_CH1 as alternate function.

Thanks

**Comment by terjeio** [comment]
I do not have a board to test with, you can try adding it? Code is [here](https://github.com/grblHAL/STM32F4xx/blob/6930c3d812504b6a951bdf8e6ec3ac0c5bc9d391/Inc/driver.h#L170-L251). You'll have to swap the debounce timer with the PWM timer.

**Reply by krishchan007** [reply]
To be fair I didn't entirely understand what you suggested (i'm kinda newbie with this stuff) , so what I did was to change the debounce timer. And guess what it worked. I am able to produce a varying pulse width signal on PE5. 

#define DEBOUNCE_TIMER_N            12

My assumptions:
As I compared the Timer 9 and 12 there wasn't much difference between them and also I noted from the alternate function pins regarding Timer12 was either not being used or assigned to stuff like temperature sensing (ADC) and SPI.

Idk whatever I did was right or wrong.
Any suggestions would be deeply appreciated.

**Reply by terjeio** [reply]
If using timer 12 for debounce the definitions should likely be:
```
#if SPINDLE_PWM_TIMER_N == 9
#define DEBOUNCE_TIMER_N            12
#define DEBOUNCE_TIMER_IRQn         TIM8_BRK_TIM12_IRQn        // !
#define DEBOUNCE_TIMER_IRQHandler   TIM8_BRK_TIM12_IRQnHandler // !
#else
#define DEBOUNCE_TIMER_N            9
#define DEBOUNCE_TIMER_IRQn         TIM1_BRK_TIM9_IRQn       // !
#define DEBOUNCE_TIMER_IRQHandler   TIM1_BRK_TIM9_IRQHandler // !
#endif
#define DEBOUNCE_TIMER              timer(DEBOUNCE_TIMER_N)
#define DEBOUNCE_TIMER_CLOCK_ENA    timerCLKENA(DEBOUNCE_TIMER_N)
```
Can you verify that the limit switches still triggers an alarm (enable hard limits with `$21=1` first).

You added something like this at the end of the PWM definitions?
```
#elif SPINDLE_PWM_PORT_BASE == GPIOE_BASE
  #if SPINDLE_PWM_PIN == 5 // PE5 - TIM9_CH1
    #define SPINDLE_PWM_TIMER_N     9
    #define SPINDLE_PWM_TIMER_CH    1
    #define SPINDLE_PWM_TIMER_INV   0
    #define SPINDLE_PWM_TIMER_AF    3
  #endif
#endif
```

**Reply by krishchan007** [reply]
Sorry for the late reply. 
I had already added the changes you mentioned in PWM definitions.
About the verification of hard limits, idk how hard limits work entirely, but after the changes you mentioned were made, once the limits are triggered, grbl stops responding, unlock commands or grbl reset doesn't work. Only after Hard resetting the board using reset button, the board responds.

**Reply by krishchan007** [reply]
I tried with the unchanged version of grbl, it throws an alarm.

**Reply by terjeio** [reply]
>  it throws an alarm.

Which one? And have you read [this](https://github.com/grblHAL/core/wiki/First-Run-Grbl-Settings)?

**Reply by krishchan007** [reply]
ALARM:1.

And I've read the guide. I inverted all the input pins including limits and control pins, for the sake of easy testing without attaching any physical switches.

**Reply by terjeio** [reply]
> once the limits are triggered, grbl stops responding, unlock commands or grbl reset doesn't work.

Ok, that could be due to the debounce timer interrupt not beeing cleared and thus firing continously. I'll check that.

ALARM:1 is hard limits triggered, what is the `?` response?

**Reply by krishchan007** [reply]
ALARM:1
[MSG:Reset to continue]
<Alarm|MPos:0.000,0.000,0.000,0.000,0.000|Bf:35,1020|FS:0,0|Pn:Z|Ov:100,100,100>
<Alarm|MPos:0.000,0.000,0.000,0.000,0.000|Bf:35,1020|FS:0,0|Pn:Z>

**Reply by terjeio** [reply]
Ok, the Z limit pin is asserted - this triggers the hard limits alarm: `|Pn:Z`

**Reply by krishchan007** [reply]
Yes. Thanks for the explanation. Looking forward to a solution for the "not responding" issue.

**Reply by terjeio** [reply]
Try switching to timer 13 for debounce in _driver.h_, there seems to be a conflict with USB for timer 12.
```
#if SPINDLE_PWM_TIMER_N == 9
#define DEBOUNCE_TIMER_N            13
#define DEBOUNCE_TIMER_IRQn         TIM8_UP_TIM13_IRQn       // !
#define DEBOUNCE_TIMER_IRQHandler   TIM8_UP_TIM13_IRQHandler // !
#else
...
```

![image](https://user-images.githubusercontent.com/20260062/193324420-eec52ea7-f41e-41b4-9c45-40d0a2f87913.png)

I did not find anything in the reference manual related to this when I did a quick search, so I decided to try timer13.

**Reply by krishchan007** [reply]
Switched to timer 13. No issues regarding hard limits or PWM output whatsoever. Greatly appreciate your help. Thank you.
